[Blueprint] Add permission validation for import steps

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Part of WOOPLUG-2482

This PR adds permission validation for Blueprint steps to ensure users have the required capabilities before executing each step. This enhancement improves security by preventing unauthorized users from performing sensitive operations.

### Key Changes:

1. Added permission checks for all Blueprint importers:
   - `ImportActivatePlugin`: Requires `activate_plugins` capability
   - `ImportActivateTheme`: Requires `switch_themes` capability
   - `ImportDeactivatePlugin`: Requires `deactivate_plugins` capability
   - `ImportDeletePlugin`: Requires both `deactivate_plugins` and `delete_plugins` capabilities
   - `ImportInstallPlugin`: Requires `install_plugins` capability
   - `ImportInstallTheme`: Requires `install_themes` capability
   - `ImportSetSiteOptions`: Requires `manage_options` capability
   - `ImportRunSql`: Requires `manage_options` and additional specific capabilities based on SQL content

2. Enhanced `ImportStep` class:
   - Added capability validation before processing steps
   - Improved validation return types for better error handling
   - Added new test case for capability validation

3. Simplified `importSchema` class to reduce duplication

4. Fix [CliResultFormatter.php](https://github.com/woocommerce/woocommerce/pull/57294/files#diff-0ebb6e1099f165f32f003774df65aa9708944e646435543c14b14421d2cbada1) to show the messages

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. pnpm install
2. Enable the Blueprint feature `wp option set woocommerce_feature_blueprint_enabled yes`
3. Create a test user with limited permissions (e.g., Editor role) `wp user create test test@example.com --role=editor`
5. Download this [blueprint.json](https://github.com/user-attachments/files/19771588/blueprint.json) file
6. Run `wp wc blueprint import blueprint.json --user=test --show-messages=all`
7. Confirm that it fails with a message indicating that the user does not have the required capabilities
8. Test with admin user and confirm that import succeeds `wp wc blueprint import blueprint.json --user=admin --show-messages=all`
9. Try with a user that has different capabilities and confirm that it fails with a message indicating that the user does not have the required capabilities

<img width="1280" alt="Screenshot 2025-04-16 at 14 18 51" src="https://github.com/user-attachments/assets/b3d02f61-c566-45cb-b823-0e9549596e14" />


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Add permission validation for import steps

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Refactor ImportSchema and ImportStep classes to enhance permissions validation

- Removed the built-in step processors from ImportSchema, simplifying the class structure.
- Introduced a new ImportStep class to handle step processing, improving modularity.
- Added capability checks in step processors to ensure users have the necessary permissions before executing steps.
- Updated validation methods to return boolean values, enhancing clarity on validation outcomes.

chihsuan committed:
Update ImportStepTest to extend WP_UnitTestCase and add capability validation test

- Changed ImportStepTest to extend WP_UnitTestCase for better integration with WordPress testing framework.
- Added a new test to validate error handling when user capabilities are insufficient for executing a step, enhancing test coverage and robustness of the import process.
- Introduced a check_step_capabilities method in DummyImporter to support capability validation.

chihsuan committed:
Fix function existence check in CliResultFormatter to use namespaced format_items function

- Updated the function existence check for format_items to use the correct namespaced path '\WP_CLI\Utils\format_items'.
- Reorganized the conditional logic to throw an exception if the function is not found before calling it, improving error handling and clarity.

chihsuan committed:
Update ImportStep to use step definition in success result

- Modified the import method to utilize the step definition's step property in the success result, enhancing clarity and ensuring the correct step context is used during the import process.

Commit:
Add changefile(s) from automation for the following project(s): packages/php/blueprint

louwie17 commented on packages/php/blueprint/src/Importers/ImportDeletePlugin.php:
```diff
@@ -44,4 +44,16 @@ public function process( $schema ): StepProcessorResult {
+	 */
+	public function check_step_capabilities( $schema ): bool {
+		return current_user_can( 'deactivate_plugins' ) &&
+		current_user_can( 'delete_plugins' );
```
Could this just be a single line?

chihsuan replied:
Sure!

chihsuan replied:
Reformatted in 17f24e5cdf2fcd9a9aacd8b52e792ab2918544f9

louwie17 commented on packages/php/blueprint/src/Importers/ImportRunSql.php:
```diff
@@ -49,4 +47,36 @@ public function process( $schema ): StepProcessorResult {
+		$sql_lowercase = strtolower( trim( $sql ) );
+
+		// Check for specific operations.
+		if ( strpos( $sql_lowercase, 'posts' ) !== false ) {
```
Both this and the `users` one below are quite generic and could easily be included in another query context that doesn't relate to the posts or users table specifically.
Should we stick to `wp_posts`, `wp_postmeta` and same for the users with `wp_users` and `wp_usermeta`?
Otherwise I may even lean to changing this so that a user has all three for any sql query: `current_user_can( 'manage_options' ) && current_user_can( 'edit_posts' ) && current_user_can( 'edit_users' )`

chihsuan replied:
I'll change the logic to check all three for any SQL query.

chihsuan replied:
Updated in 76ca635b9501260f8aaedef6835a9c35bee784a3.

louwie17 reviewed (changes requested):
Tested well, just left one small comment around formatting and another in regards to the run sql permission check, we should be more specific there. Although ideally I think we should just require all three ( basically always administrator privileges, given it is a generic SQL query ).

chihsuan committed:
Refactor capability checks in ImportRunSql to check all permissions

chihsuan committed:
Refactor capability check formatting in ImportDeletePlugin for improved readability

louwie17 reviewed (approved):
Thanks for the updates @chihsuan, this remained to test well, nice work!

chihsuan committed:
[Blueprint] Add permission validation for export steps and REST API (#57303)

* Enhance ExportSchema and related exporters with capability checks and error handling

- Updated the export method in ExportSchema to return WP_Error for invalid landing page paths instead of throwing exceptions, improving error handling.
- Added check_step_capabilities method to various StepExporter implementations to validate user permissions before executing steps, ensuring only authorized users can perform exports.
- Updated CLI export command to handle WP_Error responses gracefully, providing clear error messages to the user.
- Enhanced documentation for methods to clarify expected return types and capabilities checks.

* Refactor permission checks in RestApi for export and import actions

- Updated permission callbacks for export and import methods to use specific checks: `check_export_permission` and `check_import_permission`.
- Enhanced permission validation to ensure only users with appropriate capabilities can perform export and import actions, improving security and clarity in user permissions.

* Improve error handling in ExportSchema and RestApi

- Updated error message in ExportSchema to clarify the context of insufficient permissions during export steps.
- Added handling for WP_Error responses in RestApi to return appropriate HTTP responses, enhancing API robustness and user feedback.

* Improve error handling in slotfill.js for export functionality

- Enhanced error handling in the Blueprint component to differentiate between various error types, including string errors and specific permission-related errors.
- Added user-friendly error messages for insufficient permissions during export, improving clarity for users.
- Ensured that the export functionality is re-enabled after error handling, maintaining a smooth user experience.

* Enhance steps_payload_to_blueprint_steps method in RestApi

- Added checks to ensure that 'settings', 'plugins', and 'themes' arrays are not only set but also contain elements before processing them, improving the robustness of the steps payload handling.

* Refactor error handling in ExportSchema for consistency

- Updated the ExportSchema class to use the WP_Error class consistently without the namespace prefix, improving code readability and maintaining consistency across error handling.
- Clarified return type documentation for the export method to reflect the use of WP_Error, enhancing developer understanding of the method's behavior.

* Initialize error state in exportBlueprint function in slotfill.js

- Added a line to reset the error state to null before starting the export process, ensuring that previous errors do not affect the current export operation.

* Add changelog

* Fix tests

* Enhance DummyExporter with strict types and additional methods

- Added strict types declaration for improved type safety.
- Introduced check_step_capabilities method to validate export capabilities.
- Enhanced documentation for existing methods to clarify their purpose and return types, improving code readability and developer understanding.

* Fix lint

* [Blueprint] Implement logging for export and import operations (#57384)

* Implement logging functionality in ExportSchema and ImportStep classes

- Introduced a Logger class to handle logging of export and import steps, improving traceability and debugging.
- Enhanced ExportSchema to log the start and completion of export steps, as well as errors encountered during the process.
- Updated ImportStep to log the start and result of the import process, including success and error messages.
- Refactored error handling in both classes to utilize the new logging system, ensuring consistent logging practices across the codebase.

* Add changelog

* Fix tests

* Refactor Logger class to utilize WP functions

- Introduced UseWPFunctions trait to encapsulate WordPress function usage.
- Updated Logger class to use wp_get_current_user_id() for retrieving the current user's ID, enhancing consistency with WordPress standards.
- Minor adjustment in ImportStepTest class to ensure proper namespace usage for WP_UnitTestCase.

* Refactor logging in ExportSchema and ImportStep classes

- Updated ExportSchema and ImportStep to utilize new logging methods in the Logger class for starting, completing, and handling failures of export and import steps.
- Removed redundant logging code and centralized logging functionality in the Logger class, enhancing code clarity and maintainability.

* Remove WC_Log_Levels and WC_Logger_Interface stubs; centralize logging in tests

- Deleted the WC_Log_Levels and WC_Logger_Interface stub files to streamline the codebase.
- Introduced a new stubs.php file to provide necessary stubs for testing, ensuring compatibility and reducing redundancy.
- Updated references in the Logger class to use fully qualified class names for WC_Log_Levels, enhancing clarity and consistency.

* Fix tests

Commit:
Add changefile(s) from automation for the following project(s): packages/php/blueprint, woocommerce, woocommerce/client/admin

elazzabi commented:
I started getting some fatal errors on JN sites probably related to this. It's blocking the main WC profiler from completing:


https://github.com/user-attachments/assets/30b83d11-987d-4446-b886-74a10e99bc62

```
Uncaught Error: Failed opening required '/srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.9.0-dev-14663310349-g771b51e/vendor/woocommerce/blueprint/tests/stubs/stubs.php' (include_path='/:.')
in /srv/htdocs/wp-content/plugins/jetpack/vendor/jetpack-autoloader/class-php-autoloader.php on line 102

Call stack:

Automattic\J\A\jpf11009ded9fc4592b6a05b61ce272b3c_jetpackⓥ14_6_a_7\a\PHP_Autoloader::load_class('WC_Log_Levels')
wp-content/plugins/wc_beta_tester_live_branch_9.9.0-dev-14663310349-g771b51e/includes/class-wc-logger.php:311
WC_Logger::debug('wcadmin_settings_view', array)
wp-content/plugins/woocommerce-beta-tester/api/tracks/class-tracks-debug-log.php:66
Tracks_Debug_Log::log_event('wcadmin_settings_view', WC_Tracks_Event)
wp-content/plugins/woocommerce-beta-tester/api/tracks/class-tracks-debug-log.php:92
Tracks_Debug_Log::log_footer_pixels('')
wp-includes/class-wp-hook.php:324
WP_Hook::apply_filters(NULL, array)
wp-includes/class-wp-hook.php:348
WP_Hook::do_action(array)
wp-includes/plugin.php:517
do_action('admin_footer', '')
wp-admin/admin-footer.php:78
require_once('/srv/htdocs/__wp__/wp-admin/admin-footer.php')
wp-admin/admin.php:298```
